### PR TITLE
Workaround for message number saving

### LINKED ISSUE #961: Can't send messages in persistent group chat
With persistent groups it sometimes happens that I can't send messages in group chat I'm connected to. Unfortunately I don't know how to reproduce this bug and it happens randomly. I appear as online and I can message people, but I am not able to send messages to a group or groups. Normally if I lost connection to a group, I should just be reconnected, but here that never happens. The only way for me to be able to send messages again is to manually leave the group and join it again or restart the client.

### PR DESCRIPTION
**Pull request by zugz**
Should fix #961, at least in its most likely incarnations.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/1362)

### CODE REVIEW

File: toxcore/group.c [hunk 1]
```diff
@@ -2618,7 +2618,7 @@ static Message_Info *find_message_slot_or_reject(uint32_t message_number, uint8_
     Message_Info *i;
 
     for (i = peer->last_message_infos; i < peer->last_message_infos + peer->num_last_message_infos; ++i) {
-        if (message_number > i->message_number) {
+        if (message_number - (i->message_number + 1) <= (1 << 31)) {
```
Comment (sudden6) [inline toxcore/group.c]: This changes the condition when `message_number == i->message_number`

**Comment by sudden6** [comment]
Please add a short description how the workaround works to that commit.

**Comment by zugz** [comment]
>> @@ -2618,7 +2618,7 @@ static Message_Info *find_message_slot_or_reject(uint32_t message_number, uint8_
>     Message_Info *i;
>
>     for (i = peer->last_message_infos; i < peer->last_message_infos + peer->num_last_message_infos; ++i) {
>-        if (message_number > i->message_number) {
>+        if (message_number - (i->message_number + 1) <= (1 << 31)) {
>
>This changes the condition when `message_number == i->message_number`

How so?
m - (m + 1) == -1 == UINT32_MAX (mod 2^{32}),
and UINT32_MAX > (1 << 31).
So both the new and the old condition are false in the case 
message_number == i->message_number. Right?

**Comment by zugz** [comment]
* Saturday, 2020-01-11 at 18:09 +0000 - sudden6 <notifications@github.com>:

>Please add a short description how the workaround works to that commit.

Done. (Not so short, though)

**Comment by sudden6** [comment]
> How so?
m - (m + 1) == -1 == UINT32_MAX (mod 2^{32}),
and UINT32_MAX > (1 << 31).
So both the new and the old condition are false in the case
message_number == i->message_number. Right?

You are right, I didn't calculate this correctly.

But what about this case:

```
message_number = 0;
i->message_number = UINT32_MAX;
// before
message_number > i->message_number // false
// now
message_number - (i->message_number + 1) <= (1 << 31)
0 - (UINT32_MAX + 1) <= (1 << 31)
// UINT32_MAX + 1 overflows to 0
0 - (0) <= (1<<31) // true
```

To fix that you would need to change the condition to:

```
((uint64_t)message_number) - (message_number2 + ((uint64_t)1)) <= (((uint64_t)1) << 31);
```
But then the case
```
message_number = 0;
i->message_number = UINT32_MAX;
``
is broken.

I think the correct behaviour depends on what this code is supposed to do.

**Comment by sudden6** [comment]
GCC complains that you're shifting into the sign bit with `(1 << 31)`, it should be `(UINT32_C(1) << 31)` to fix that.

**Comment by zugz** [comment]
* Wednesday, 2020-01-22 at 12:06 -0800 - sudden6 <notifications@github.com>:

>But what about this case:
>
>message_number = 0;
>i->message_number = UINT32_MAX;

Yes, that's the point of this commit -- in this situation, 
message_number should be considered newer than i->message_number.

**Comment by zugz** [comment]
* Wednesday, 2020-01-22 at 20:08 +0000 - sudden6 <notifications@github.com>:

>GCC complains that you're shifting into the sign bit with `(1 << 31)`, 
>it should be `(UINT32_C(1) << 31)` to fix that.

Argh. Thanks. Should be fixed now.

**Comment by sudden6** [comment]
> Yes, that's the point of this commit -- in this situation,
message_number should be considered newer than i->message_number.

Ok, then this makes sense.

Can you rebase on master? this should make the tsan test pass too.

**Pull request merged**